peregrine/resources/submission/graphql/node.py

```diff
@@ -788,3 +786,97 @@ def get_fields():
 
 
 NodeField = graphene.List(Node, args=get_node_interface_args())
+
+
+# ======================================================================
+# DataNode
+
+
+class DataNode(graphene.Interface):
+    id = graphene.ID()
+    shared_fields = None
+
+
+# return a dictionary containing the fields shared by all data nodes
+def get_shared_fields_dict():
+    if not DataNode.shared_fields:
+        shd_fields = None
+        for node in dictionary.schema:
+            schema = dictionary.schema[node]
+            if schema['category'].endswith('_file'):
+                fields = schema['properties'].keys()
+                if shd_fields is None:
+                    shd_fields = set(fields)
+                else:
+                    shd_fields = shd_fields.intersection(fields)
+        shd_fields_dict = {}
+        for field in shd_fields:
+            if field == 'file_size':
+                shd_fields_dict[field] = graphene.Int()
+            # elif field.endswith('_id'):
+            #     shd_fields_dict[field] = graphene.ID()
+            else:
+                shd_fields_dict[field] = graphene.String()
+        DataNode.shared_fields = shd_fields_dict
+    return DataNode.shared_fields
+
+
+def resolve_datanode(self, info, **args):
+    """The root query for the :class:`DataNode` node interface.
+
+    :returns:
+        A list of graphene object classes.
+
+    """
+
+    # get the list of categories that are data categories
+    data_types = filter(lambda node: dictionary.schema[node]['category'].endswith('_file'), dictionary.schema)
+
+    # get the subclasses for the data categories
+    data_types = filter(lambda node: node.label in data_types, psqlgraph.Node.get_subclasses())
```

Comment:
codacy complains about this not being a list comprehension; the comprehension would be:
```python
data_types = [
    node
    for node in psqlgraph.Node.get_subclasses()
    if node.label in data_types
]
```
or if you prefer the filter you can tell codacy to ignore it.

Also for clarity I would name this differently from the previous `data_types`—maybe rename the one above this to `data_type_labels` or something